From 99badc1665e2c820451f7bbc72fbdaab6e3efa90 Mon Sep 17 00:00:00 2001
From: Hongxu Jia <hongxu.jia@windriver.com>
Date: Wed, 20 Jun 2018 10:22:09 +0800
Subject: [PATCH 3/7] tweak regex of ostree-system-generator

The ostree has boot params ostree="/ostree/1" on grub
and uboot which is a synlink point to
the dir contains "boot.0/OSTREE_OSNAME/***".

So tweak regex of ostree-system-generator.

Upstream-Status: Inappropriate [LAT specific]

Signed-off-by: Hongxu Jia <hongxu.jia@windriver.com>

Rebased on 20230607 for ostree 2023.3.

Signed-off-by: Kai Kang <kai.kang@windriver.com>
---
 src/libostree/ostree-impl-system-generator.c | 11 ++++++++---
 1 file changed, 8 insertions(+), 3 deletions(-)

diff --git a/src/libostree/ostree-impl-system-generator.c b/src/libostree/ostree-impl-system-generator.c
index b5a7cde2..6fc22db3 100644
--- a/src/libostree/ostree-impl-system-generator.c
+++ b/src/libostree/ostree-impl-system-generator.c
@@ -95,16 +95,21 @@ stateroot_from_ostree_cmdline (const char *ostree_cmdline, GError **error)
 {
   static GRegex *regex;
   static gsize regex_initialized;
+  char ostree_cmdline_new[PATH_MAX] = {0};
+
+  if (readlink(ostree_cmdline, ostree_cmdline_new, sizeof(ostree_cmdline_new)) < 0)
+      return glnx_null_throw (error, "Failed to readlink %s", ostree_cmdline);
+
   if (g_once_init_enter (&regex_initialized))
     {
-      regex = g_regex_new ("^/ostree/boot.[01]/([^/]+)/", 0, 0, NULL);
+      regex = g_regex_new ("boot.[01]/[^/]+/([^/]+)/", 0, 0, NULL);
       g_assert (regex);
       g_once_init_leave (&regex_initialized, 1);
     }
 
   g_autoptr (GMatchInfo) match = NULL;
-  if (!g_regex_match (regex, ostree_cmdline, 0, &match))
-    return glnx_null_throw (error, "Failed to parse %s", ostree_cmdline);
+  if (!g_regex_match (regex, ostree_cmdline_new, 0, &match))
+    return glnx_null_throw (error, "Failed to parse %s", ostree_cmdline_new);
 
   return g_match_info_fetch (match, 1);
 }
-- 
2.27.0

